### ISSUE TITLE
Clarify samples count between -e cpu and -e itimer

### ISSUE THREAD

**Issue by franz1981** [jvm-profiling-tools/async-profiler#272]
Hi,

thanks for this awesome profiler!
I cannot found any info on the doc re how to interpret samples count while using `-e cpu` and `-e itimer`. I wouldn't like it to became a blog about how sampling profilers (should) work, but I admit that writing 2 lines of doc there could help to understand the difference in samples count a user can expect depending on the context.
We can start a discussion here and I can send a PR to add it, if is welcome :+1: 

thanks,
Franz

**Comment by apangin** [comment]
Hi Franz,

Thank you for your interest in async-profiler and its internals.

Both `cpu` and `itimer` mode measure the CPU time spent by the running threads. For example, if an application uses 2 cpu cores, each with 30% utilization, and the sampling interval is 10ms, then the profiler will collect about 2 * 0.3 * 100 = 60 samples per second.

In other words, 1 profiling sample means that one CPU was actively running for N nanoseconds, where N is the profiling interval.

 - `itimer` mode is based on [`setitimer(ITIMER_PROF)`](http://man7.org/linux/man-pages/man2/setitimer.2.html) syscall, which [ideally] generates a signal every given interval of the CPU time consumed by the process.
 - `cpu` mode relies on PMU-based sampling capabilities (like Intel PEBS or AMD IBS). The idea is the same - to generate a signal every `N` nanoseconds of CPU time, which in this case is achieved by configuring PMU to generate an interrupt every `K` CPU cyles.

**Comment by apangin** [comment]
So, in ideal world, both `itimer` and `cpu` should collect the same number of samples. Typically the profiles indeed look very similar.

[In some cases](https://github.com/golang/go/issues/14434) `cpu` profile appears a bit more accurate though, since the signal is delivered exactly to the thread that overflowed a hardware counter.

**Comment by franz1981** [comment]
Thanks @apangin!

I've written a simple experiment to understand this:
```java
public class SamplingExperiment {

    public static void main(String[] args) {
        final long targetNanos = TimeUnit.MILLISECONDS.toNanos(10);
        while (true) {
            spinWork(targetNanos);
            //I can measure latency of spinWork and use to sleep too
            sleep(targetNanos);
        }
    }

    private static void sleep(long nanos) {
        long remaining = nanos;
        while (remaining > 0) {
            final long start = System.nanoTime();
            LockSupport.parkNanos(remaining);
            final long elapsed = System.nanoTime() - start;
            remaining = nanos - elapsed;
        }
    }

    private static void spinWork(long nanos) {
        final long start = System.nanoTime();
        while (System.nanoTime() - start <= nanos) {

        }
    }
}
```
And I'm getting the same number of samples for `itimer` and `cpu` ie 493-494 that's expected, given that this should use 50% of a single core. 
Obviously this code can work only for not too small `targetNanos`, but will try on a bigger machine to see hot it behaves there too.

**Comment by apangin** [comment]
I assume the question is answered, and I may close the issue.

**Comment by petrbouda** [comment]
Hi guys, it's a very interesting topic. If I understand correctly.

CPU mode: Every CPU calculates Process Time spent on CPU of all running threads and if the `interval` is reached then SIGPROF is propagated by the currently running thread to the process's signal-handler where hands over a stacktrace (PERF_SAMPLE_CALLCHAIN) with KERNEL and USER frames and call AGCT to match USER frames with Java/Native/JVM frames. Am I correct?

Does mean that all threads open perf_event_open into a single group?

`    
if (args._ring == RING_USER) {
        attr.exclude_kernel = 1;
} else if (args._ring == RING_KERNEL) {
        attr.exclude_user = 1;
} else if (!Symbols::haveKernelSymbols()) {
        Profiler::_instance.updateSymbols(true);
        attr.exclude_kernel = Symbols::haveKernelSymbols() ? 0 : 1;
}
`
Does this mean that if I am not allowed to register receiving KERNEL stacks (because of paranoid, cap_sys_admin) I will receive at least USER frames to match it inside SIGPROF handler after 10ms on a single CPU (depending on an interval). However, busy-client from this example should still be discovered (https://github.com/apangin/codeone2019-java-profiling), Am I right?

Thanks a lot and great job in terms of AsyncProfiler! :)

**Comment by franz1981** [comment]
hi @petrbouda 

If I run the profiler with `itimer`I get
![image](https://user-images.githubusercontent.com/13125299/98774098-968bdd00-23ea-11eb-9e3d-e79073be22dc.png)

While with `cpu`
![image](https://user-images.githubusercontent.com/13125299/98774120-a4416280-23ea-11eb-9daa-1f2e552c35c8.png)

The amount of sample is a bit different too (for 5 seconds of profiling time):

- `itimer`: 801 samples
- `cpu`: 1787 samples

I hope that answer your question

**Comment by petrbouda** [comment]
I did absolutely the same test. However, the difference is significant I think. It's why I would like to understand CPU mode more, to figure it out how it works not only in this case but to be able to reason about it in general.

**Comment by franz1981** [comment]
I believe the CPU mode is just relying on AGCT to check who is on stack and that means that we don't have kernel info if the native method is doing something for real or not..Not sure, I should check the code there @apangin knows it for sure

**Comment by petrbouda** [comment]
I think that SIGPROF signal determines who is currently on CPU (the guy who delivers the signal to the registered handler inside the process) and AGCT is just a tool to get a java stacktrace for the given thread (without a need to be at a safepoint) and then match the java stacktrace with USER part of stacktrace delivered by the thread. 
This is what how I got understood when I was looking at AsyncProfiler's source-code, but I'm a very weak C-code reader :)

**Comment by franz1981** [comment]
@petrbouda 
Yep, but AGCT  doesn't have any mean to tell what's the native code (on top of the perceived stack) is doing or real, if consuming real cpu time or not

**Comment by franz1981** [comment]
Discovered myself (through https://github.com/jvm-profiling-tools/async-profiler/discussions/466#discussioncomment-1394187) on Linux man pages:
```
ITIMER_PROF
              This timer counts down against the total (i.e., both user
              and system) CPU time consumed by the process.  (The
              measurement includes CPU time consumed by all threads in
              the process.)  At each expiration, a SIGPROF signal is
              generated.
```
the part re `by the process` is key ie its not "per-thread" as perf does.

But sthil the match isn't clear to me yet, indeed on https://github.com/jvm-profiling-tools/async-profiler/issues/272#issuecomment-725218272:
801 samples of itimer on 5 seconds of profiling at 100hz -> means max total 500 samples, now is possible is getting more then it?

@petrbouda if you have understood how itimer samples works

**Comment by franz1981** [comment]
It looks that using setitimer can suffer from some biasing and the current existing tests accept a 20% variance in samples (!!): see https://github.com/google/gvisor/blob/master/test/syscalls/linux/itimer.cc#L240
I'll go deeper to understand what it means

**Comment by franz1981** [comment]
See an old comment by @felixge on https://groups.google.com/g/golang-dev/c/LCJyvL90xv8

> That being said, it's easy to get confused about this. E.g. if you have two CPU intense threads, and a single setitimer(ITIMER_PROF, ...), you'll see signals arrive at twice the rate of the configured interval. That's because you end up utilizing 2s of CPU time for every 1s of wall clock time.

It's stating that I *can* get more then the expected number of samples based on the number of active (on cpu) threads in the process, but I cannot see evidence on kernel code about it yet

**Comment by petrbouda** [comment]
> Discovered myself (through [#466 (reply in thread)](https://github.com/jvm-profiling-tools/async-profiler/discussions/466#discussioncomment-1394187)) on Linux man pages:
> 
> ```
> ITIMER_PROF
>               This timer counts down against the total (i.e., both user
>               and system) CPU time consumed by the process.  (The
>               measurement includes CPU time consumed by all threads in
>               the process.)  At each expiration, a SIGPROF signal is
>               generated.
> ```
> 
> the part re `by the process` is key ie its not "per-thread" as perf does.
> 
> But sthil the match isn't clear to me yet, indeed on [#272 (comment)](https://github.com/jvm-profiling-tools/async-profiler/issues/272#issuecomment-725218272): 801 samples of itimer on 5 seconds of profiling at 100hz -> means max total 500 samples, how is possible is getting more then it?
> 
> @petrbouda if you have understood how itimer samples works

I have no idea :) ... currently, I use AsyncProfiler inside Openshift. I always need to upload it to the container, execute and download the result. Unfortunately, our infra guys don't want to touch perf_event variables
```
sysctl kernel.perf_event_paranoid
sysctl kernel.kptr_restrict
```
It would be a security incident, they said :) ... At least, I can enjoy Java frames and after the long fight JVM's frames as well, I enforced  JDK with symbols inside containers :)

**Comment by franz1981** [comment]
I've solved it @petrbouda 

try compile this
```c
#define _GNU_SOURCE
#include <stdio.h>
#include <stdlib.h>
#include <signal.h>
#include <pthread.h>
#include <sys/time.h>
	
	static void myHandler(int iSig)
	{ 
		printf("In myHandler with argument %d\n", iSig);
	}

	void* foo(void* p){
  		// Print value received as argument:
  		for (;;) {
		
		}
	}

int main(void)
{ 
	struct itimerval sTimer;
	signal(SIGPROF, myHandler);
	sTimer.it_value.tv_sec = 1; /* Send first signal in 1 second */
	sTimer.it_value.tv_usec = 0; /* and 0 microseconds. */
	sTimer.it_interval.tv_sec = 1; /* Send subsequent signals in 1 sec */
	sTimer.it_interval.tv_usec = 0; /* and 0 microsecond intervals. */
	setitimer(ITIMER_PROF, &sTimer, NULL);
	printf("Entering an infinite loop\n");
	pthread_t id;

  	int j = 1;
  	pthread_create(&id, NULL, foo, &j);
  	for (;;) {
		
	}
	return 0; /* Never get here. */
}
```
with
```bash
$ gcc -Wall testtimer.c -lrt -o testtimer
```
You will see that the SIGPROF handler is running twice as the "expected" rate (by me, and wrongly assumed now, I see).
It means that it's being fired according to the cpu time used by the whole process: here I'm fully using 2 threads (in busy loop)
so the CPU time spent is happening at twice the speed as by having a single thread doing it.

This is going to solve my problem with the number of samples mentioned on https://github.com/jvm-profiling-tools/async-profiler/issues/272#issuecomment-1092777444 but probably we're getting less then the one using perf because we're not considering cpu time spent only on kernel stack(s) - but it shouldn't be like that TBH, given that doc state
```
              This timer counts down against the total (i.e., both user
              and system) CPU time consumed by the process.
```
*both user and system* -> maybe it's async-profiler, given that it doesn't collect kernel stack traces it's missing to report samples that SIGPROF has correctly fired?

@apangin knows

**Comment by felixge** [comment]
@franz1981 you summoned me, so here are my two cents :)

If I skimmed this thread correctly, one issue is that setitimer sometimes delivers less samples than the perf events based cpu mode?

If yes, we observed the same issue in Go. It was fixed in 1.18 by using timer_create on Linux. See this article for details: https://felixge.de/2022/02/11/profiling-improvements-in-go-1.18/

Hope that helps?

**Comment by franz1981** [comment]
Wow @felixge yes yes and yes. It knows totally makes sense ☺️ many thanks and sorry for the ping!
